File: apollo-api/src/main/java/com/apollographql/apollo/api/internal/SimpleResponseWriter.java
package com.apollographql.apollo.api.internal;

import com.apollographql.apollo.api.ResponseField;
import com.apollographql.apollo.api.ResponseFieldMarshaller;
import com.apollographql.apollo.api.ResponseWriter;
import com.apollographql.apollo.api.ScalarType;
import com.apollographql.apollo.api.internal.json.JsonWriter;
import com.apollographql.apollo.response.CustomTypeAdapter;
import com.apollographql.apollo.response.CustomTypeValue;
import com.apollographql.apollo.response.ScalarTypeAdapters;
import okio.Buffer;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static com.apollographql.apollo.api.internal.Utils.checkNotNull;
import static com.apollographql.apollo.api.internal.json.Utils.writeToJson;

public class SimpleResponseWriter implements ResponseWriter {
  private final Map<String, Object> data;
  private final ScalarTypeAdapters scalarTypeAdapters;

  public SimpleResponseWriter(@NotNull ScalarTypeAdapters scalarTypeAdapters) {
    this.data = new LinkedHashMap<>();
    this.scalarTypeAdapters = checkNotNull(scalarTypeAdapters, "scalarTypeAdapters == null");
  }

  public String toJson(String indent) throws IOException {
    final Buffer buffer = new Buffer();

    final JsonWriter jsonWriter = JsonWriter.of(buffer);
    jsonWriter.setIndent(indent);
    jsonWriter.beginObject();
    jsonWriter.name("data");

    writeToJson(data, jsonWriter);

    jsonWriter.endObject();
    jsonWriter.close();

    return buffer.readUtf8();
  }

  @Override public void writeString(@NotNull ResponseField field, @Nullable String value) {
    data.put(field.responseName(), value);
  }

  @Override public void writeInt(@NotNull ResponseField field, @Nullable Integer value) {
    data.put(field.responseName(), value);
  }

  @Override public void writeLong(@NotNull ResponseField field, @Nullable Long value) {
    data.put(field.responseName(), value);
  }

  @Override public void writeDouble(@NotNull ResponseField field, @Nullable Double value) {
    data.put(field.responseName(), value);
  }

  @Override public void writeBoolean(@NotNull ResponseField field, @Nullable Boolean value) {
    data.put(field.responseName(), value);
  }

  @SuppressWarnings("unchecked")
  @Override public void writeCustom(@NotNull ResponseField.CustomTypeField field, @Nullable Object value) {
    if (value == null) {
      data.put(field.responseName(), null);
    } else {
      final CustomTypeAdapter typeAdapter = scalarTypeAdapters.adapterFor(field.scalarType());
      final CustomTypeValue customTypeValue = typeAdapter.encode(value);
      data.put(field.responseName(), customTypeValue.value);
    }
  }

  @Override public void writeObject(@NotNull ResponseField field, @Nullable ResponseFieldMarshaller marshaller) {
    if (marshaller == null) {
      data.put(field.responseName(), null);
    } else {
      SimpleResponseWriter objectResponseWriter = new SimpleResponseWriter(scalarTypeAdapters);
      marshaller.marshal(objectResponseWriter);
      data.put(field.responseName(), objectResponseWriter.data);
    }
  }

  @Override public void writeFragment(@Nullable ResponseFieldMarshaller marshaller) {
    if (marshaller != null) {
      marshaller.marshal(this);
    }
  }

  @Override public <T> void writeList(@NotNull ResponseField field, @Nullable List<T> values, @NotNull ListWriter<T> listWriter) {
    if (values == null) {
      data.put(field.responseName(), null);
    } else {
      CustomListItemWriter listItemWriter = new CustomListItemWriter(scalarTypeAdapters);
      listWriter.write(values, listItemWriter);
      data.put(field.responseName(), listItemWriter.data);
    }
  }

  private static class CustomListItemWriter implements ListItemWriter {
    private final List<Object> data;
    private final ScalarTypeAdapters scalarTypeAdapters;

    CustomListItemWriter(ScalarTypeAdapters scalarTypeAdapters) {
      this.data = new ArrayList<>();
      this.scalarTypeAdapters = scalarTypeAdapters;
    }

    @Override public void writeString(@Nullable String value) {
      data.add(value);
    }

    @Override public void writeInt(@Nullable Integer value) {
      data.add(value);
    }

    @Override public void writeLong(@Nullable Long value) {
      data.add(value);
    }

    @Override public void writeDouble(@Nullable Double value) {
      data.add(value);
    }

    @Override public void writeBoolean(@Nullable Boolean value) {
      data.add(value);
    }

    @SuppressWarnings("unchecked")
    @Override public void writeCustom(@NotNull ScalarType scalarType, @Nullable Object value) {
      if (value == null) {
        data.add(null);
      } else {
        final CustomTypeAdapter typeAdapter = scalarTypeAdapters.adapterFor(scalarType);
        final CustomTypeValue customTypeValue = typeAdapter.encode(value);
        data.add(customTypeValue.value);
      }
    }

    @Override public void writeObject(@Nullable ResponseFieldMarshaller marshaller) {
      if (marshaller == null) {
        data.add(null);
      } else {
        SimpleResponseWriter objectResponseWriter = new SimpleResponseWriter(scalarTypeAdapters);
        marshaller.marshal(objectResponseWriter);
        data.add(objectResponseWriter.data);
      }
    }

    @Override public <T> void writeList(@Nullable List<T> items, @NotNull ListWriter<T> listWriter) {
      if (items == null) {
        data.add(null);
      } else {
        CustomListItemWriter listItemWriter = new CustomListItemWriter(scalarTypeAdapters);
        listWriter.write(items, listItemWriter);
        data.add(listItemWriter.data);
      }
    }
  }
}

File: apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/internal/SimpleResponseWriter.kt
package com.apollographql.apollo.api.internal

import com.apollographql.apollo.api.ResponseField
import com.apollographql.apollo.api.CustomScalar
import com.apollographql.apollo.api.CustomScalarAdapters
import com.apollographql.apollo.api.ResponseField.Companion.leafType
import com.apollographql.apollo.api.internal.json.JsonWriter
import com.apollographql.apollo.api.internal.json.Utils.writeToJson
import com.apollographql.apollo.api.internal.json.use
import okio.Buffer
import okio.IOException

class SimpleResponseWriter(private val customScalarAdapters: CustomScalarAdapters) : ResponseWriter {
  private val data: MutableMap<String, Any?> = LinkedHashMap()

  @Throws(IOException::class)
  fun toJson(indent: String?): String {
    return Buffer().apply {
      JsonWriter.of(this).use { jsonWriter ->
        jsonWriter.indent = indent
        jsonWriter.beginObject()
        jsonWriter.name("data")
        writeToJson(data, jsonWriter)
        jsonWriter.endObject()
      }
    }.readUtf8()
  }

  override fun writeString(field: ResponseField, value: String?) {
    data[field.responseName] = value
  }

  override fun writeInt(field: ResponseField, value: Int?) {
    data[field.responseName] = value
  }

  override fun writeLong(field: ResponseField, value: Long?) {
    data[field.responseName] = value
  }

  override fun writeDouble(field: ResponseField, value: Double?) {
    data[field.responseName] = value
  }

  override fun writeBoolean(field: ResponseField, value: Boolean?) {
    data[field.responseName] = value
  }

  override fun writeCustom(field: ResponseField, value: Any?) {
    if (value == null) {
      data[field.responseName] = null
    } else {
      val typeAdapter = customScalarAdapters.adapterFor<Any>(field.type.leafType())
      val jsonElement = typeAdapter.encode(value)
      data[field.responseName] = jsonElement.toRawValue()
    }
  }

  override fun writeObject(field: ResponseField, block: ((ResponseWriter) -> Unit)?) {
    if (block == null) {
      data[field.responseName] = null
    } else {
      val objectResponseWriter = SimpleResponseWriter(customScalarAdapters)
      block.invoke(objectResponseWriter)
      data[field.responseName] = objectResponseWriter.data
    }
  }

  override fun <T: Any> writeList(
      field: ResponseField,
      values: List<T?>?,
      block: (item: T, listItemWriter: ResponseWriter.ListItemWriter) -> Unit
  ) {
    if (values == null) {
      data[field.responseName] = null
    } else {
      val listItemWriter = CustomListItemWriter(customScalarAdapters)
      values.forEach {
        if (it == null) {
          listItemWriter.writeNull()
        } else {
          block(it, listItemWriter)
        }
      }
      data[field.responseName] = listItemWriter.data
    }
  }

  private class CustomListItemWriter(private val customScalarAdapters: CustomScalarAdapters) : ResponseWriter.ListItemWriter {
    val data = ArrayList<Any?>()

    override fun writeString(value: String) {
      data.add(value)
    }

    override fun writeInt(value: Int) {
      data.add(value)
    }

    override fun writeLong(value: Long) {
      data.add(value)
    }

    override fun writeDouble(value: Double) {
      data.add(value)
    }

    override fun writeBoolean(value: Boolean) {
      data.add(value)
    }

    override fun writeCustom(customScalar: CustomScalar, value: Any) {
      val typeAdapter = customScalarAdapters.adapterFor<Any>(customScalar)
      val jsonElement = typeAdapter.encode(value)
      data.add(jsonElement.toRawValue())
    }

    override fun writeObject(block: ((ResponseWriter) -> Unit)) {
      val objectResponseWriter = SimpleResponseWriter(customScalarAdapters)
      block.invoke(objectResponseWriter)
      data.add(objectResponseWriter.data)
    }

    override fun <T: Any> writeList(items: List<T?>, block: (item: T, listItemWriter: ResponseWriter.ListItemWriter) -> Unit) {
      val listItemWriter = CustomListItemWriter(customScalarAdapters)
      items.forEach {
        if (it == null) {
          listItemWriter.writeNull()
        } else {
          block(it, listItemWriter)
        }
      }
      data.add(listItemWriter.data)
    }

    fun writeNull() {
      data.add(null)
    }
  }
}
